llpc/lower/llpcSpirvProcessGpuRtLibrary.cpp

```diff
@@ -349,6 +351,107 @@ void SpirvProcessGpuRtLibrary::createConvertF32toF16WithRoundingMode(Function *f
+  auto argIt = func->arg_begin();
+  Value *roundMode = m_builder->CreateLoad(int32Ty, argIt++);
+  Value *operation = m_builder->CreateLoad(int32Ty, argIt++);
+  Value *src0 = m_builder->CreateLoad(retType, argIt++);
+  Value *src1 = m_builder->CreateLoad(retType, argIt);
+
+  // Save the current roundMode.
+  Value *tmpRoundMode = m_builder->CreateNamedCall(
+      "llvm.amdgcn.s.getreg", IntegerType::get(m_builder->getContext(), 32), {m_builder->getInt32(hwRegField)}, {});
+
+  auto rmDefault = BasicBlock::Create(*m_context, ".rmDefault", func, nullptr);
+  auto opBlock = BasicBlock::Create(*m_context, ".opBlock", func, nullptr);
+  auto opDefault = BasicBlock::Create(*m_context, ".opDefault", func, nullptr);
+  auto rmSwitch = m_builder->CreateSwitch(roundMode, rmDefault);
+
+  for (uint32_t rm = static_cast<uint32_t>(RoundingMode::TowardZero);
+       rm <= static_cast<uint32_t>(RoundingMode::TowardNegative); rm++) {
+    std::string blockName = ".roundMode" + std::to_string(rm);
+    auto roundBlock = BasicBlock::Create(*m_context, blockName, func, rmDefault);
+    m_builder->SetInsertPoint(roundBlock);
+    rmSwitch->addCase(m_builder->getInt32(rm), roundBlock);
+    m_builder->CreateNamedCall("llvm.amdgcn.s.setreg", Type::getVoidTy(m_builder->getContext()),
+                               {m_builder->getInt32(hwRegField), m_builder->getInt32(rm)}, {});
+    m_builder->CreateBr(opBlock);
+  }
+  m_builder->SetInsertPoint(rmDefault);
+  m_builder->CreateBr(opBlock);
+
+  m_builder->SetInsertPoint(opBlock);
+  auto opSwitch = m_builder->CreateSwitch(operation, opDefault);
+  for (uint32_t op = OperationType::Add; op <= OperationType::Mul; op++) {
+    std::string blockName = ".opMode" + std::to_string(op);
+    auto opMode = BasicBlock::Create(*m_context, blockName, func, opDefault);
+    m_builder->SetInsertPoint(opMode);
+    opSwitch->addCase(m_builder->getInt32(op), opMode);
+    Value *result = PoisonValue::get(retType);
+    if (op == OperationType::Add) {
+      result = m_builder->CreateFAdd(src0, src1);
+      // Restore the roundMode.
+      m_builder->CreateNamedCall("llvm.amdgcn.s.setreg", Type::getVoidTy(m_builder->getContext()),
+                                 {m_builder->getInt32(hwRegField), tmpRoundMode}, {});
+      m_builder->CreateRet(result);
+    } else if (op == OperationType::Sub) {
+      result = m_builder->CreateFSub(src0, src1);
+      // Restore the roundMode.
+      m_builder->CreateNamedCall("llvm.amdgcn.s.setreg", Type::getVoidTy(m_builder->getContext()),
+                                 {m_builder->getInt32(hwRegField), tmpRoundMode}, {});
+      m_builder->CreateRet(result);
+    } else {
+      result = m_builder->CreateFMul(src0, src1);
+      // Restore the roundMode.
+      m_builder->CreateNamedCall("llvm.amdgcn.s.setreg", Type::getVoidTy(m_builder->getContext()),
+                                 {m_builder->getInt32(hwRegField), tmpRoundMode}, {});
+      m_builder->CreateRet(result);
```

Comment:
The setreg + return suffix should be moved out of the if/else chain since it's logically redundant.